### DISCUSSION TITLE
Customize map pins in .NET MAUI

### DISCUSSION THREAD

**Discussion by giscus** [VladislavAntonyuk/vladislavantonyuk.github.io#73] [Announcements]
# Customize map pins in .NET MAUI

Customize map pins in .NET MAUI.

https://vladislavantonyuk.github.io/articles/Customize-map-pins-in-.NET-MAUI/

**Comment by VinShen** [comment]
Good work Vladislav Antonyuk.

I tried the custom image on the Map Pins. It works but if Pin count is 120 or more, the Map loading takes lot of time on Android. It's in Minutes.
iOS is little better.
What can we increase Map loading with more custom Image Pins?
Please suggest. Thanks

**Reply by VladislavAntonyuk** [reply]
It is limited by phone memory. You can try not add pin if it is not in the area.

**Reply by VinShen** [reply]
Thanks Vladislav Antonyuk for your response.

We have about 150 bike stations in a city which we have to show together.

Our current App (in Stores) in Xamarin Forms App using MapRenderer, GoogleMap.IInfoWindowAdapter with Xamarin.FormsMaps for Android works very well.
(On same phones)

I am working on Maui (.Net 8) App for the same. All other functionalities are ported and working.
But now stuck up with Custom Map Pins for Android (extremely slow in refreshing on same Android phones)
iOS is good and fast.

It will be great if you can suggest anything in this regards.

Thanks

**Reply by VladislavAntonyuk** [reply]
just a question, have you used the code from the article or the app sample? I made some tweaks in handler in the app sample. Could you try that code?

**Reply by VinShen** [reply]
Hi Vladislav Antonyuk.

I did try the App Sample from the latest code.
The Marker Pin loading speed of is much better with the latest changes in CustomMapHandler made by you. Thanks for the same.
Map view is good with even 150+ pins. 

But in ViewModel implementation the MarkerClick Event is not firing.

I used Observable Collections of CustomPin as itemsource for Map.
With Map in MainPage.xaml, Pin.MarkerClicked does fire in MainPage.xaml.cs for all CustomPins properly.

But for MainPageViewModel implementation, when clicked on Marker, the Pin.MarkerClicked is not fired for the CustomPins.

When clicked on Marker, the CustomMarkerClickListener -> OnMarkerClick method does execute and send SendMarkerClick() to Pin, but the Pin.MarkerClicked event is not fired for the CustomPins.

Testing with .Net 8 and latest packages.
Please note that with earlier implementation of CustomMapHandler (without pinsToAdd, pinsToRemove, ...) the MainPageViewModel Marker Click was working properly.

Can you please give some suggestions in this regard.

Thanks

**Reply by VladislavAntonyuk** [reply]
We add event listeners here: https://github.com/VladislavAntonyuk/MauiSamples/blob/main/MauiMaps/Platforms/Android/MapCallbackHandler.cs#L18

and look for the pin here:
https://github.com/VladislavAntonyuk/MauiSamples/blob/main/MauiMaps/Platforms/Android/CustomMarkerClickListener.cs#L17

You can debug and check if pin is found on click

**Reply by VinShen** [reply]
Hi Vladislav Antonyuk.

I confirmed public bool OnMarkerClick(Android.Gms.Maps.Model.Marker marker) of CustomMarkerClickListener.cs does fire and Proper Pin is found at L17.

When I use 
public ObservableCollection<CustomPin> MyPins { get; } = new();
at MainPage.xaml.cs the 
var myPin = new CustomPin();
MyMap.Pins.Add(myPin);
myPin.MarkerClicked += async (s, e) =>
                {
                    e.HideInfoWindow = true;
                    Pin pin = (Pin)s;
                    string pinName = pin.Label;
                    await Toast.Make($"{pinName}").Show();
                };
does work properly.

But when I use
public ObservableCollection<CustomPin> MyPins { get; } = new();
at MainPageViewModel.cs and bind it with xaml map as ItemsSource="{Binding MyPins}"

var myPin = new CustomPin();
MyPins.Add(myPin)
myPin.MarkerClicked += async (s, e) => .....

does not work.

Please note that with earlier implementation of CustomMapHandler (without pinsToAdd, pinsToRemove, ...) this (MainPageViewModel implementation) was also working properly.

Am I missing something?

**Reply by VinShen** [reply]
Hi Vladislav Antonyuk.

Sorry !!

It was my mistake. I was using map from my own class CMap derived from Microsoft.Maui.Controls.Maps.Map

I used the CMap instead of IMap in CustomMapHandler and MapCallbackHandler.

Now Map, CustomPins, Click Events every thing is working fine and fast even for 200+ pins for Android.

Thank you for the code and support. You saved my App. Great job !!

**Reply by OpenSoftsrl** [reply]
Hi VinShen, can you share your CMap class?

**Reply by Ducvu1996** [reply]
I have the same problem, my project have 1000 pins, but it takes a lot of time, if I dont use handle for IOS, it is quite fast, but i need to handle to show the marker in Pin on IOS, because the default pin show the marker under, I need it show in the top, any one have suggestion for me, thank you so much.

**Comment by Suhailbsth** [comment]
i have tried this code. but the "MauiMKMapView" shows error. is there any nuget package to be installed for this to work?

**Reply by VladislavAntonyuk** [reply]
No, make sure you use it in iOS file.

**Comment by bdgza** [comment]
Hi, trying to port maps with custom pins form XF to MAUI with .NET 8. When I copy your Android code for CustomMapHandler I get an error on the line:
```var mapReady = new MapCallbackHandler(this);```
because `MapCallbackHandler` is inaccessible. Perhaps this is something that changed with .NET 8. I am not able to find what the alternative is supposed to be.

**Reply by VladislavAntonyuk** [reply]
you are looking for this file: https://github.com/VladislavAntonyuk/MauiSamples/blob/35981ed61bd831c9cdb9c60ef7e3d21688df7a99/MauiMaps/Platforms/Android/MapCallbackHandler.cs

**Reply by bdgza** [reply]
Thank you!

**Comment by Brosten** [comment]
Thanks for sharing!
Anyone has a working example for Windows? Is it even possible, since it's using webView+BingMaps?

**Reply by VladislavAntonyuk** [reply]
WinUI 1.5 sdk brings native platform instead of webview. Most likely MAUI team will implement it in .NET 9

**Comment by GJungstedt** [comment]
Thank you! Is it okay if i use this code for my own project?

**Reply by VladislavAntonyuk** [reply]
Yes. If it’s open source, I would appreciate if you leave a link to the repo. Thank you

**Reply by GJungstedt** [reply]
Thank you again, here is the repo https://github.com/GurraBern/PloggingApp

**Comment by Ducvu1996** [comment]
Hi Vladislav Antonyuk. I tried the custom image on the Map Pins. My project have 1000 pins,  I have no problem with Xamarin but when I change to MAUI, the Map loading takes lot of time on both IOS and Android. 
please suggest. Thanks.

**Reply by VladislavAntonyuk** [reply]
Hello, the suggestion is adding only markers that are visible on the screen area. I am not sure if all 1000 pins are useful for the user.

I hope one day MAUI team will add Pin Icon API, because now we recreate each pin multiple times.

**Comment by NKJRao** [comment]
Hi,
This map handler is based on .net 6, when we use this in 8 or 9 then there are lot's of big changes is there can anyone provide map handler for latest one please.

Nishant

**Reply by pagirre** [reply]
What do you mean? Which aspects of CustomMapHandler would you update?

**Reply by Ducvu1996** [reply]
I am using Maps in .NET 8. It is ok if you just handler for about 100 pins load in once.

**Reply by pagirre** [reply]
My problem is as follows: I am using .net8 and I want to load dynamically 1000 pins approximately. I have already coded it. But it is unusable. Anyone any suggestion? 

Based on the comment from @NKJRao  I thought the improvements/updates expected would help in performance but there has been no response.

**Comment by ncson8920** [comment]
it's nice! but how to I customize InfoWindow of Pin?

**Reply by Ducvu1996** [reply]
It is easy, I have done, if you need, contact me

**Reply by ncson8920** [reply]
yes, please! I try to customize InfoWindow in few day but it stay in 0

**Reply by Ducvu1996** [reply]
johnducvu1996@gmail.com this is my email.

**Comment by Midnayt** [comment]
Nice!
What about changing pin image dynamically?
for example, last pin must be one image and other pins - with different. and when new position is added, change pin source image for new pin and previous?

**Reply by ncson8920** [reply]
with customPin, you add IsLast property default true . When you add new Pin, search the last Pin and update IsLast property to false and after add new Pin. In AddPins function of CustomMapHandler, check if IsLast is true SetIcon image1 if false SetIcon image2. Hope it can help you!

**Reply by Midnayt** [reply]
problem in CustomMapHandler - how I can change icon for previous pin? if it is already exist?

**Reply by VladislavAntonyuk** [reply]
You need to remove and add a pin with a new image. Pin is immutable

**Reply by Ducvu1996** [reply]
is it possible if I custom annotation?, I want the annotation show in the top instead of below like default. [VladislavAntonyuk] please.

**Reply by VladislavAntonyuk** [reply]
I don’t remember the specific setting for that. It displays annotation in the free space.

**Comment by GenCodeInc** [comment]
Good afternoon Vladislav,

When I call SendMarkerClick();; the InfoWindow does not pop up, Im tryin to fix it but so far no luck :(

    private async void OnClickAirportButtonClicked(object sender, EventArgs e)
    {
        foreach (var pin in MainMap.Pins.ToList())
        {
            if (pin is CustomPin)
            {
                if (pin.Label.Contains("Miami"))
                {
                    ((CustomPin)pin).SendMarkerClick();
                    break;
                }
            }
        }
    }

**Reply by VladislavAntonyuk** [reply]
Hello @GenCodeInc, could you please provide a small repro sample?
For now make sure SendMarkerClick is called at all.

**Comment by GenCodeInc** [comment]
sure, made this public, click the button at top it goes to Miami but the InfoWindow does not popup.
https://github.com/GenCodeInc/Airports

**Comment by prashanth742** [comment]
Hi The custom pins are working fine but when using this, The direction and map links are not showing up. Is there a wayto add them back?

**Comment by AngeloLazzari** [comment]
Hello, great post Vladislav! 
We  are now using your code in our app, that is great! Thank you very much for share your knowhow. 

We would like to change the pin in runtime, when a user clicks a pin it has to change the imagesource, how can we do it in IOS and ANDROID? 

Actually we tried to use the MarkerClicked method of the pin changing runtime the imagesource but the pin is not change the image. 

thank you very much

**Reply by VladislavAntonyuk** [reply]
Hello, @AngeloLazzari ! Thank you for the feedback. It is not possible on Android. The only way to change pin image is recreate the pin. As for iOS, I don’t remember, but it should work.

**Reply by AngeloLazzari** [reply]
ok thank you I will investigate and I will update you about how it will go.. thank you!!!!

**Comment by PauchardThomas** [comment]
Hi,

on iOS when I customize a Pin with an ImageSource, all the others are no longer displayed. Any idea ?

**Reply by PauchardThomas** [reply]
My bad, justed updated this line : if (pin is CustomPin cp && cp.ImageSource != null) and it's working fine now.

**Reply by AngeloLazzari** [reply]
Great job man!

**Comment by rajeshangappan** [comment]
How can i create MAP for travel. I have start and end point location i need a route info in my app.

**Reply by VladislavAntonyuk** [reply]
for now it is only available in native app: https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/map?view=net-maui-9.0#launch-the-map-app-with-directions
or you can manually draw it using a polyline: https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/map?view=net-maui-9.0#create-a-polyline
or you can try this sdk: https://developers.arcgis.com/net/maui/sample-code/find-route/

**Comment by jaroslavic** [comment]
On iOS, adding many polylines to the map seems to cause freezing. Is there any way to improve this, or perhaps an alternative component we could consider?

**Reply by VladislavAntonyuk** [reply]
It's a bad practice to add all pins to the map; a better approach is to analyze the visible area and add only the necessary pins there. Also, you can limit the zoom feature.
Alternatively, you can see how I implemented it for my app: https://github.com/VladislavAntonyuk/WorldExplorer/tree/main/src/Client/Client/Controls/WorldExplorerMap
